### DISCUSSION TITLE
Welcome to Why-Not-Image-Carousel Discussions!

### DISCUSSION THREAD

**Discussion by ImaginativeShohag** [ImaginativeShohag/Why-Not-Image-Carousel#31] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by rama-2402** [comment]
Hi @ImaginativeShohag I'm new to Android Development, I trying out a bunch of new stuff to learn where i came across this awesome library. !!! This makes life a lot easier. Very clear documentation and easy implementation. Thank you for that. 

I used this library inside a recycler view to get something like netflix or amazon prime video homepage, and it works perfectly. But the problem is it is not registering the any item clicks. Is there any solution for that or any right way to implement this library inside a recycler view?

**Reply by ImaginativeShohag** [reply]
If you provide a demo project then I can debug the problem. Upload a demo project with the specific problem in GitHub and share the repo here. Then I can quickly look at it and can give you suggestions.

**Comment by NevoSaporta** [comment]
Hi @ImaginativeShohag, i'm an Android & iOS developer at a startup called Clarifruit. We're currently using your awesome library in our Android app. 

I'm trying to get a reference of the carousel's default next & previous buttons (without setting custom views to them using the app:nextButtonId/app:previousButtonId) and i wanted to ask you what is the best method of doing so?

**Reply by ImaginativeShohag** [reply]
You can get default previous/next button reference like this:

```kotlin
val prevBtn = binding.carousel.findViewById<MaterialButton>(binding.carousel2.previousButtonId)
val nextBtn = binding.carousel.findViewById<MaterialButton>(binding.carousel2.nextButtonId)
```

**Reply by NevoSaporta** [reply]
Cool! Thanks for you help

**Comment by SamKotBrown** [comment]
Hello @ImaginativeShohag, I'm new to kotlin developping and i wanted to know if i can change the style of the "dots" on the control that tell me in what item i am at that moment. I saw that in the CircleIndicator git page there is another style with lines that flip around when changing the item.

Is there a way i can change the style on your image viewer ?

Thank you !

**Reply by ImaginativeShohag** [reply]
Yes you can customize `CircleIndicator2`. You can add your custom `CircleIndicator2` by using `setIndicator()` method like this:

```kotlin
val customIndicator: CircleIndicator2 = findViewById(R.id.custom_indicator)
carousel.setIndicator(customIndicator)
```

Or you can `getIndicator()` and customize the existing indicator programmatically.

**Comment by Florent-Breton** [comment]
Hi, 
I implemented your library with success and it's really cool. But how to open the carousel (or slide to) a specific position without doing carousel.next() several times?
As I used an infinite carousel, I tried   
            `carousel.currentVirtualPosition = pos
            Log.e("TEST",  carousel.currentVirtualPosition.toString())`
or also carousel.currentPosition = pos, but the log always display -1

Thanks for your help

NB: in opposite to addData, it could be cool to have a removeData method

**Reply by ImaginativeShohag** [reply]
You can use `carousel.currentPosition = desirePosition` to scroll to that position. It should work. If it's not working, give me some sample code please, so that I can check it.

And for `removeData` method, you can remove the item from the original dataset then use `setData` method to replace the data. Noted the `removeData` feature request. :)

**Reply by Florent-Breton** [reply]
Here's the code, I don't know what is wrong
`binding.carousel.registerLifecycle(lifecycle)

        viewModel.setPhotos(
            intent.extras?.getParcelableArrayList<Photo>(Constants.PHOTO) as ArrayList<Photo>,
            fromLastToFirst
        )

        viewModel.photos.observe(this, {
            val list = mutableListOf<CarouselItem>()
            viewModel.photos.value!!.forEach { photo ->
                list.add(
                    CarouselItem(
                        imageUrl = photo.uri?.path
                    )
                )
            }

            binding.carousel.setData(list)
        })

        binding.carousel.carouselListener = object : CarouselListener {
            override fun onCreateViewHolder(
                layoutInflater: LayoutInflater,
                parent: ViewGroup
            ): ViewBinding {
                return PhotoSliderItemBinding.inflate(layoutInflater, parent, false)
            }

            override fun onBindViewHolder(binding: ViewBinding, item: CarouselItem, position: Int) {
                val currentBinding = binding as PhotoSliderItemBinding

                currentBinding.photoSliderItem.apply {
                    val photo = File(viewModel.photos.value!![position].uri?.path!!)
                    Picasso.get()
                        .load(photo)
                        .into(this)
                }
            }
        }

        binding.carousel.onScrollListener = object : CarouselOnScrollListener {
            override fun onScrollStateChanged(
                recyclerView: RecyclerView,
                newState: Int,
                position: Int,
                carouselItem: CarouselItem?
            ) {
                super.onScrollStateChanged(recyclerView, newState, position, carouselItem)
                viewModel.setPosition(position)
            }

            override fun onScrolled(
                recyclerView: RecyclerView,
                dx: Int,
                dy: Int,
                position: Int,
                carouselItem: CarouselItem?
            ) {
                super.onScrolled(recyclerView, dx, dy, position, carouselItem)
                this@PhotoViewerActivity.dx = dx
                this@PhotoViewerActivity.position = position
            }
        }

        val pos = intent.extras?.getInt(Constants.POSITION, 0)
        Log.e("TEST", pos!!.toString())
        if (pos != 0) {
            binding.carousel.currentPosition = pos
        }`

**Reply by ImaginativeShohag** [reply]
Move the following part:

```kotlin
    val pos = intent.extras?.getInt(Constants.POSITION, 0)
    Log.e("TEST", pos!!.toString())
    if (pos != 0) {
        binding.carousel.currentPosition = pos
    }
```

Inside `viewModel.photos.observe()` after `binding.carousel.setData(list)`.

This should work. Let me know.

**Reply by Florent-Breton** [reply]
Yes, that works, thanks.
